### ISSUE TITLE
Lints `(a - b) < f32::EPSILON` in favor of `(a - b).abs() < f32::EPSILON`

### ISSUE THREAD

**Issue by kangalioo** [rust-lang/rust-clippy#5819]
### What it does

Lints comparisons of float against `f32::EPSILON` if `.abs()` is not being used.

Note: in this text I'm using the f32 type to make my point, but of course this lint should also apply to f64 and any other floating point type that we might get in the future

### Categories

- Kind: `clippy::correctness`

### What is the advantage of the recommended code over the original code

- The original code without `.abs()` has 99% probably a bug.

### Drawbacks
There is a case where this code might be intended behavior: when the user can ensure that b is larger than a.

However,
1. this is probably a niche case
2. it is very easy to miss this very important condition, and very hard to debug later if the condition has been accidentally violated (i.e. if b turns out smaller than a at some point in the program). So it can't hurt to explicitly alert the reader of the code using `#[allow(clippy::float_equality_without_abs)]`

### Example
```rust
pub fn is_roughly_equal(a: f32, b: f32) -> bool {
    (a - b) < f32::EPSILON
}
```

should be written as:

```rust
pub fn is_roughly_equal(a: f32, b: f32) -> bool {
    (a - b).abs() < f32::EPSILON
}
```

**Comment by kangalioo** [comment]
Do the Clippy developers accept this as a lint suggestion, or are there objections?

I'm asking because I'm thinking about attempting a pull request, but I don't want to start working on it if the feature won't even be accepted

**Comment by flip1995** [comment]
This is a good idea, IMO. I also don't see a reason why someone shouldn't add `.abs()` here. I don't think it has any performance implications either. If the `abs` call was left out for performance reasons, it is probably better to state this explicitly in the code, which then can be done with an `allow`. 

So I'd say go ahead and implement this. There's a WIP document for [Clippy-hacking-basics](https://github.com/flip1995/rust-clippy/blob/basics/doc/basics.md) and a document for [how to write a new lint](https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md). If you have any questions, feel free to ask here, in discord or open a WIP PR.

**Comment by 1c3t3a** [comment]
I would like to work on this, unless @kangalioo did already start!

**Comment by kangalioo** [comment]
I did start, though I gave up because it was so frustrating.

Here's my work so far, you can most probably reuse the test suite. The code.. maybe. Not sure how good the actual lint code or if it should be reworked from scratch https://github.com/kangalioo/rust-clippy/commit/086269b1325ed4fa2bc2494c7adfc7ee0c8df076

**Comment by 1c3t3a** [comment]
I started implementing the lint and opened a pull request. Thanks for the snippets @kangalioo!

**Comment by kangalioo** [comment]
Well this was quick :p Nice to see this finished, thank you

**Comment by 1c3t3a** [comment]
No problem, thank you for all the input! Didn't think that this gets merged so fast neither :D

**Comment by kangalioo** [comment]
By the way, I skimmed over the code changes in the PR and I could see for example this bit:
```rust

let suggestion = if lhs_string.starts_with('(') {
    format!("{}.abs()", lhs_string)
} else {
    format!("({}).abs()", lhs_string)
};
```
If I recall correctly, this was imported from my code. I thought this was a horrible hack but I didn't know how to fix it, and this part of the reason I got frustrated, but apparently this was fine after all.

Now I'm wondering what you actually changed and added to the code, and which parts were already ok

**Comment by 1c3t3a** [comment]
I actually took two parts from your code that my version didn't have at that point. One was the suggestion and the other one was logic around whether there is a `Gt` with the subtraction on the right side or the `Lt` with the subtraction on the left side. I just modified it to use a match instead of an `if .. else if .. else` construction:

```rust
if let ExprKind::Binary(op, ref left, ref right) = expr.kind {
            match op.node {
                BinOpKind::Lt => {
                    lhs = left;
                    rhs = right;
                },
                BinOpKind::Gt => {
                    lhs = right;
                    rhs = left;
                },
                _ => return,
            };
} else {
    return;
}
```
I wrote the rest of the program on my own, but it ended up pretty close to your solution. I think the only thing you didn't include was the type-check of the float variables in the subtraction! I hope this answers your question.

**Comment by kangalioo** [comment]
Yeah it does answer my question, thank you :)
